File: typescript/Array.ts
class FixedArray<T> {
    private items: T[];
    private readonly capacity: number;
    private size: number;
    constructor(capacity: number) {
        if (capacity <= 0) {
            throw new Error('Capacity must be a positive integer.');
        }
        this.capacity = capacity;
        this.items = new Array(capacity);
        this.size = 0;
    }

    private checkIndex(index: number) {
        if (index < 0 || index >= this.size) {
            throw new Error(`Index out of bounds: ${index}. Current size: ${this.size}.`);
        }
    }

    private checkCapacity() {
        if (this.size >= this.capacity) {
            throw new Error('Fixed Array is full.Cannot add more elements.');
        }
    }

    /**
     * @description 从尾部插入
     */
    public append(element: T) {
        this.checkCapacity();
        this.items[this.size] = element;
        this.size++;
    }

    /**
     * @description 从头部插入
     */
    public prepend(element: T) {
        this.checkCapacity();
        this.addAt(0, element);
    }

    /**
     * @description 从第几位添加一个元素
     */
    public addAt(index: number, element: T) {
        if (index < 0 || index > this.size) {
            throw new Error(
                `Insertion index is out of bounds. Valid range: 0 to ${this.size}. Requested: ${index}`
            );
        }
        this.checkCapacity();
        // 从index开始往后移动一位
        for (let i = this.size - 1; i >= index; i--) {
            this.items[i + 1] = this.items[i];
        }
        this.items[index] = element;
        this.size++;
    }

    /**
     * 从第几位删除一个元素
     */
    public remove(index: number): T {
        this.checkIndex(index);
        const removedElement = this.items[index];
        for (let i = index; i < this.size; i++) {
            this.items[i] = this.items[i + 1];
        }
        this.size--;
        return removedElement;
    }

    public popFront() {
        this.remove(0);
    }

    public popBack() {
        this.remove(this.size - 1);
    }

    public get(index: number) {
        this.checkIndex(index);
        return this.items[index] as (undefined | T);
    }

    public set(index: number, element: T) {
        this.checkIndex(index);
        this.items[index] = element;
    }

    public getFirst() {
        return this.get(0) as (undefined | T);
    }

    public getLast() {
        return this.get(this.size - 1) as (undefined | T);
    }

    public sort(comparator?: (a: T, b: T) => number) {
        if (this.size <= 1) return;
        const cmp =
            comparator ??
            ((a: any, b: any) => {
                if (a < b) return -1;
                if (a > b) return 1;
                return 0;
            });

        const arr = this.items;
        const quickSort = (left: number, right: number) => {
            if (left >= right) return;

            const pivotIndex = partition(left, right);
            quickSort(left, pivotIndex - 1);
            quickSort(pivotIndex + 1, right);
        };
        const partition = (left: number, right: number): number => {
            const pivot = arr[right];
            let i = left;

            for (let j = left; j < right; j++) {
                if (cmp(arr[j], pivot) < 0) {
                    [arr[i], arr[j]] = [arr[j], arr[i]];
                    i++;
                }
            }

            [arr[i], arr[right]] = [arr[right], arr[i]];
            return i;
        };
        quickSort(0, this.size - 1);
    }

    public getSize() {
        return this.size;
    }

    public getCapacity(){
        return this.capacity;
    }

    public getElements(){
        return this.items;
    }

    public toArray(){
        return this.items.slice(0, this.size);
    }

    public toString(){
        return `[Size: ${this.size}, Capacity: ${this.capacity}, ${this.toArray().join()}]`
    }
}

class DynamicArray<T> {
    private items: T[];
    private size: number;
    private growthFactor = 2;
    constructor(initialCapacity: number = 10) {
        if (initialCapacity <= 0) {
            throw new Error('Capacity must be a positive integer.');
        }
        this.items = new Array(initialCapacity);
        this.size = 0;
    }

    private ensureCapacity(){
        if (this.size === this.items.length) {
            const newItems: T[] = new Array(this.size * this.growthFactor);
            for (let i = 0; i < this.items.length; i++) {
                newItems[i] = this.items[i];
            }
            this.items = newItems;
        }
    }
    private checkIndex(index: number) {
        if (index < 0 || index >= this.size) {
            throw new Error(`Index out of bounds: ${index}. Current size: ${this.size}.`);
        }
    }

    /**
     * @description 从尾部插入
     */
    public append(element: T) {
        this.ensureCapacity();
        this.items[this.size] = element;
        this.size++;
    }

    /**
     * @description 从头部插入
     */
    public prepend(element: T) {
        this.addAt(0, element);
    }

    /**
     * @description 从第几位添加一个元素
     */
    public addAt(index: number, element: T) {
        if (index < 0 || index > this.size) {
            throw new Error(
                `Insertion index is out of bounds. Valid range: 0 to ${this.size}. Requested: ${index}`
            );
        }
        this.ensureCapacity();
        // 从index开始往后移动一位
        for (let i = this.size - 2; i > index; i--) {
            this.items[i + 1] = this.items[i];
        }
        this.items[index] = element;
        this.size++;
    }

    /**
     * 从第几位删除一个元素
     */
    public remove(index: number): T {
        this.checkIndex(index);
        const removedElement = this.items[index];
        for (let i = index; i < this.size; i++) {
            this.items[i] = this.items[i + 1];
        }
        this.size--;
        return removedElement;
    }

    public popFront() {
        this.remove(0);
    }

    public popBack() {
        this.remove(this.size - 1);
    }

    public get(index: number) {
        this.checkIndex(index);
        return this.items[index];
    }

    public set(index: number, element: T) {
        this.checkIndex(index);
        this.items[index] = element;
    }

    public getFirst() {
        return this.get(0);
    }

    public getLast() {
        return this.get(this.size - 1);
    }

    public sort(comparator?: (a: T, b: T) => number) {
        if (this.size <= 1) return;
        const cmp =
            comparator ??
            ((a: any, b: any) => {
                if (a < b) return -1;
                if (a > b) return 1;
                return 0;
            });

        const arr = this.items;
        const quickSort = (left: number, right: number) => {
            if (left >= right) return;

            const pivotIndex = partition(left, right);
            quickSort(left, pivotIndex - 1);
            quickSort(pivotIndex + 1, right);
        };
        const partition = (left: number, right: number): number => {
            const pivot = arr[right];
            let i = left;

            for (let j = left; j < right; j++) {
                if (cmp(arr[j], pivot) < 0) {
                    [arr[i], arr[j]] = [arr[j], arr[i]];
                    i++;
                }
            }

            [arr[i], arr[right]] = [arr[right], arr[i]];
            return i;
        };
        quickSort(0, this.size - 1);
    }

    public getSize() {
        return this.size;
    }

    public getElements(){
        return this.items;
    }

    public toArray(){
        return this.items.slice(0, this.size);
    }

    public toString(){
        return `[Size: ${this.size}, ${this.toArray().join()}]`
    }
}

const fixedArray = new FixedArray(10);
for (let i = 0; i < 8; i++) {
    fixedArray.append(i);
}
console.log(fixedArray.toString());
fixedArray.addAt(3, 11);
console.log(fixedArray.toString());
fixedArray.remove(3);
console.log(fixedArray.toString());

File: go/src/data-structures/array.go
package datastructures

import (
	"errors"
	"fmt"
)

type FixedArray[T any] struct {
	size     int
	capacity int
	items    []T
}

func NewFixedArray[T any](capacity int) (*FixedArray[T], error) {
	if capacity < 0 {
		return nil, errors.New("Capacity must be a positive integer.")
	}
	return &FixedArray[T]{
		size:     0,
		capacity: capacity,
		items:    make([]T, capacity),
	}, nil
}

func (f *FixedArray[T]) checkIndex(index int) error {
	if index >= f.size {
		return fmt.Errorf("Index out of bounds: %d. Current size: %d.", index, f.size)
	}
	return nil
}

func (f *FixedArray[T]) checkCapacity() error {
	if f.size >= f.capacity {
		return errors.New("Fixed Array is full.Cannot add more elements.")
	}
	return nil
}

func (f *FixedArray[T]) Append(element T) error {
	err := f.checkCapacity()
	if err != nil {
		return nil
	}
	f.items[f.size] = element
	f.size++
	return nil
}

func (f *FixedArray[T]) AddAt(index int, element T) error {
	if index < 0 || index > int(f.size) {
		return fmt.Errorf("Insertion index is out of bounds. Valid range: 0 to %d. Requested: %d", f.size, index)
	}
	for i := f.size - 1; i >= index; i++ {
		f.items[i+1] = f.items[i]
	}
	f.items[index] = element
	f.size++
	return nil
}

func (f *FixedArray[T]) Prepend(element T) error {
	err := f.AddAt(0, element)
	if err != nil {
		return err
	}
	return nil
}

func (f *FixedArray[T]) Remove(index int) (T, error) {
	err := f.checkIndex(index)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	removedElement := f.items[index]
	for i := index; i < f.size; i++ {
		f.items[i] = f.items[i+1]
	}
	f.size--

	return removedElement, nil
}

func (f *FixedArray[T]) PopFront() error {
	_, err := f.Remove(0)
	if err != nil {
		return err
	}
	return nil
}

func (f *FixedArray[T]) PopBack() error {
	_, err := f.Remove(f.size - 1)
	if err != nil {
		return err
	}
	return nil
}

func (f *FixedArray[T]) Get(index int) (T, error) {
	err := f.checkIndex(index)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	el := f.items[index]
	return el, nil
}

func (f *FixedArray[T]) Set(index int, element T) error {
	err := f.checkIndex(index)
	if err != nil {
		return err
	}
	f.items[index] = element
	return nil
}

func (f *FixedArray[T]) GetFirst() (T, error) {
	el, err := f.Get(0)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	return el, nil
}

func (f *FixedArray[T]) GetLast() (T, error) {
	el, err := f.Get(f.size - 1)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	return el, nil
}

func (f *FixedArray[T]) GetSize() int {
	return f.size
}

func (f *FixedArray[T]) GetCapacity() int {
	return f.capacity
}

func (f *FixedArray[T]) GetElements() []T {
	return f.items
}

func (f *FixedArray[T]) ToSlice() []T {
	return f.items[:f.size]
}

func (f *FixedArray[T]) ToString() string {
	var s string
	for _, el := range f.items {
		s = fmt.Sprintf("%s, %s", s, el)
	}
	return fmt.Sprintf("[Size: %d, Capacity: %d, %s]", f.size, f.capacity, s)
}

type DynamicArray[T any] struct {
	items          []T
	size           int
	capacityFactor int
}

func NewDynamicArray[T any](initialCapacity int) (*DynamicArray[T], error) {
	if initialCapacity < 0 {
		return nil, errors.New("Capacity must be a positive integer.")
	}
	return &DynamicArray[T]{
		items:          make([]T, initialCapacity),
		size:           0,
		capacityFactor: 2,
	}, nil
}

func (d *DynamicArray[T]) checkIndex(index int) error {
	if index >= d.size {
		return fmt.Errorf("Index out of bounds: %d. Current size: %d.", index, d.size)
	}
	return nil
}

func (d *DynamicArray[T]) ensureCapacity() {
	if d.size == len(d.items) {
		newCapacity := d.capacityFactor * d.size
		if newCapacity < 1 {
			newCapacity = 1
		}
		newItems := make([]T, newCapacity)
		// for index, el := range d.items {
		// 	newItems[index] = el
		// }
		copy(newItems, d.items)
	}
}

func (d *DynamicArray[T]) Append(element T) {
	d.ensureCapacity()
	d.items[d.size] = element
	d.size++
}

func (d *DynamicArray[T]) AddAt(index int, element T) error {
	if index < 0 || index > int(d.size) {
		return fmt.Errorf("Insertion index is out of bounds. Valid range: 0 to %d. Requested: %d", d.size, index)
	}
	d.ensureCapacity()
	for i := d.size - 1; i >= index; i++ {
		d.items[i+1] = d.items[i]
	}
	d.items[index] = element
	d.size++
	return nil
}

func (d *DynamicArray[T]) Prepend(element T) error {
	err := d.AddAt(0, element)
	if err != nil {
		return err
	}
	return nil
}

func (d *DynamicArray[T]) Remove(index int) (T, error) {
	err := d.checkIndex(index)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	removedElement := d.items[index]
	for i := index; i < d.size; i++ {
		d.items[i] = d.items[i+1]
	}
	d.size--

	return removedElement, nil
}

func (d *DynamicArray[T]) PopFront() error {
	_, err := d.Remove(0)
	if err != nil {
		return err
	}
	return nil
}

func (d *DynamicArray[T]) PopBack() error {
	_, err := d.Remove(d.size - 1)
	if err != nil {
		return err
	}
	return nil
}

func (d *DynamicArray[T]) Get(index int) (T, error) {
	err := d.checkIndex(index)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	el := d.items[index]
	return el, nil
}

func (d *DynamicArray[T]) Set(index int, element T) error {
	err := d.checkIndex(index)
	if err != nil {
		return err
	}
	d.items[index] = element
	return nil
}

func (d *DynamicArray[T]) GetFirst() (T, error) {
	el, err := d.Get(0)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	return el, nil
}

func (d *DynamicArray[T]) GetLast() (T, error) {
	el, err := d.Get(d.size - 1)
	if err != nil {
		var zeroValue T
		return zeroValue, err
	}
	return el, nil
}

func (d *DynamicArray[T]) GetSize() int {
	return d.size
}
func (d *DynamicArray[T]) GetElements() []T {
	return d.items
}

func (d *DynamicArray[T]) ToSlice() []T {
	return d.items[:d.size]
}

func (d *DynamicArray[T]) ToString() string {
	var s string
	for _, el := range d.items {
		s = fmt.Sprintf("%s, %s", s, el)
	}
	return fmt.Sprintf("[Size: %d, %s]", d.size, s)
}
